kphp_polyfills.php

```diff
@@ -769,56 +769,122 @@ function memory_get_static_usage(): int {
 
 #endregion
 
-#region job workers
 
-interface KphpJobWorkerRequest {}
-interface KphpJobWorkerResponse {}
+#region job workers
 
-class DummyKphpJobWorkerRequest implements KphpJobWorkerRequest {}
+// contains declarations for KPHP job workers api
+// it has NO PHP IMPLEMENTATION; kphp_job_worker_start() isn't supposed to be called in plain PHP
 
 /**
- * @param KphpJobWorkerRequest $request
- * @return int
+ * Job workers are separate processes which can be invoked from HTTP workers to parallelize your PHP code.
+ * Unlike forks, that are coroutines actually, job workers can parallelize CPU-consuming execution.
+ * @see kphp_job_worker_start accepting \KphpJobWorkerRequest and returning future<\KphpJobWorkerResponse>|false
+ * @see wait accepting future<T>|false and returning ?T
+ * To handle job requests in an entrypoint, test for $_SERVER["JOB_ID"]
  */
-function kphp_job_worker_start(KphpJobWorkerRequest $request): int {
-  critical_error("not implemented");
-  return 0;
+interface KphpJobWorkerRequest {
 }
 
 /**
- * @param int   $job_id
- * @param float $tmp_wait_timeout
- * @return KphpJobWorkerRequest
+ * Job workers are separate processes which can be invoked from HTTP workers to parallelize your PHP code.
+ * @see \KphpJobWorkerRequest
  */
-function kphp_job_worker_wait(int $job_id, float $tmp_wait_timeout = -1): KphpJobWorkerRequest {
-  critical_error("not implemented");
-  return new DummyKphpJobWorkerRequest();
+interface KphpJobWorkerResponse {
 }
 
 /**
- * @return KphpJobWorkerRequest
+ * Classes marked with this interface can be shared across multiple job workers (for reading only due to immutability).
+ * This is used for optimization: it allows copying an object only once while launching multiple job workers.
+ * Consider the docs for details.
+ * @see kphp_job_worker_start_multi
+ * @kphp-immutable-class
  */
-function kphp_job_worker_fetch_request(): KphpJobWorkerRequest {
-  critical_error("not implemented");
-  return new DummyKphpJobWorkerRequest();
+interface KphpJobWorkerSharedMemoryPiece {
 }
 
 /**
- * @param KphpJobWorkerResponse $response
+ * When a job worker fails, this built-in instance is returned from wait().
+ * In other words, it's an error result for future<KphpJobWorkerResponse>
+ * Hence, having called wait(), you should test the result with instanceof, handling an error and probably providing a local execution path.
  */
-function kphp_job_worker_store_response(KphpJobWorkerResponse $response) {
-  critical_error("not implemented");
+class KphpJobWorkerResponseError implements KphpJobWorkerResponse {
+  public function getError(): string {
+    return '';
+  }
+
+  public function getErrorCode(): int {
+    // JOB_MEMORY_LIMIT_ERROR = -101
```

Comment:
It would be nice to turn these comments into constants of the class.

Reply:
This should be done in KPHP at first. And only then they could be exposed to polyfills. 
So, let's leave them as comments for now